### PR TITLE
Make Build.exec more concurrent

### PR DESCRIPTION
**Pull request by rgrinberg**
Both/Seq/Map2 can all be done concurrently

### CODE REVIEW

**Comment by rgrinberg** [comment]
I added a commit to make `Build.all` a primitive. This allows us to use `Fiber.parallel_map` which is much faster than iterating over a list with `fork`.

**Comment by rgrinberg** [comment]
And I optimized some other `Build.With_targets` primitives. I forgot to do this when I introduced `Build.{seq,both}` initially.

File: src/dune_engine/static_deps.ml [hunk 1]
```diff
@@ -21,3 +21,9 @@ let union x y =
 
 let paths { action_deps; rule_deps } =
   Dep.Set.paths (Dep.Set.union action_deps rule_deps)
+
+let union_map xs ~f =
+  let sets = List.map xs ~f in
+  { rule_deps = Dep.Set.union_map sets ~f:(fun x -> x.rule_deps)
+  ; action_deps = Dep.Set.union_map sets ~f:(fun x -> x.action_deps)
```
Comment (snowleopard) [inline src/dune_engine/static_deps.ml:28]: It seems a bit strange to compute the union twice, discarding half of the output each time. Wouldn't it be faster to do it only once, followed by two `map`s?

**Comment by snowleopard** [comment]
This looks good to me but I have a hard time convincing myself that concurrent execution of these primitives doesn't change the semantics. I guess this relies on some `Fiber` laws but which ones? 

@rgrinberg Do you see a simple way to justify these transformations?

File: src/dune_engine/static_deps.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_engine/static_deps.ml:28]: Hmm, but we're not computing the union twice. The only computation we're repeating is folding over `sets` twice. Which I suppose we could avoid by folding. I'll do that.

**Comment by rgrinberg** [comment]
I think it does change the semantics, but any program that observes the difference was likely ill advised. I'm sure we could write some weird program of the form `Build.both (Build.fiber ..) (Build.fiber ..)` where the 2nd fiber somehow depended on the first one to finish, but it would be a mistake to do that.

**Comment by snowleopard** [comment]
> I think it does change the semantics, but any program that observes the difference was likely ill advised. I'm sure we could write some weird program of the form `Build.both (Build.fiber ..) (Build.fiber ..)` where the 2nd fiber somehow depended on the first one to finish, but it would be a mistake to do that.

@rgrinberg I agree that such cases should be considered bugs, but how do we know that Dune doesn't have such bugs right now? By switching to using `parallel_map` when evaluating such build expressions, we can unmask such existing bugs.

Of course, we do get some confidence from the test suite that says everything is good. I wonder, however, if we could somehow detect problematic build expressions and error out instead of potentially producing incorrect build result.

**Review by snowleopard — APPROVED** [review]
The PR looks good, however @rgrinberg says that it changes the semantics of the `Build` language and the difference in behaviours can be observed if Dune has some incorrect `Build` expressions (where parallel branches secretly depend on each other). There is therefore a chance that we might need to do some follow-up bug fixing in Dune.

**Comment by rgrinberg** [comment]
> I agree that such cases should be considered bugs, but how do we know that Dune doesn't have such bugs right now? By switching to using parallel_map when evaluating such build expressions, we can unmask such existing bugs.

We don't use `Build.fiber` or `Build.dyn_fiber` currently apart from experimental branches, so it seems like we are pretty safe for now.

**Pull request merged**